File: src/scala/src/main/scala/Units.scala
package org.openworm.trackercommons

import kse.jsonal._
import kse.jsonal.JsonConverters._

package units {
  trait Convert extends AsJson {
    def name: String
    def toInternal(x: Double): Double
    def toExternal(x: Double): Double
    def isIdentity: Boolean = false

    def json = Json.Str(name)

    def changeToInternal(ds: Array[Double]) { var i = 0; while (i < ds.length) { ds(i) = toInternal(ds(i)); i +=1 } }
    def changeToInternal(fs: Array[Float]) { var i = 0; while (i < fs.length) { fs(i) = toInternal(fs(i)).toFloat; i += 1 } }
    def changeToInternal(dss: Array[Array[Double]]) { var i = 0; while (i < dss.length) { changeToInternal(dss(i)); i += 1 } }
    def changeToInternal(fss: Array[Array[Float]]) { var i = 0; while (i < fss.length) { changeToInternal(fss(i)); i += 1 } }

    def changeToExternal(ds: Array[Double]) { var i = 0; while (i < ds.length) { ds(i) = toExternal(ds(i)); i +=1 } }
    def changeToExternal(fs: Array[Float]) { var i = 0; while (i < fs.length) { fs(i) = toExternal(fs(i)).toFloat; i += 1 } }
    def changeToExternal(dss: Array[Array[Double]]) { var i = 0; while (i < dss.length) { changeToExternal(dss(i)); i += 1 } }
    def changeToExternal(fss: Array[Array[Float]]) { var i = 0; while (i < fss.length) { changeToExternal(fss(i)); i += 1 } }
  }

  case class IdentityConvert(name: String) extends Convert {
    def toInternal(x: Double) = x
    def toExternal(x: Double) = x
    override def isIdentity = true
  }

  case class LinearConvert(name: String, one: Double, dDim: Int, tDim: Int, aDim: Int) extends Convert {
    import LinearConvert._
    def toInternal(x: Double) = x * one
    def toExternal(x: Double) = x / one
    override def isIdentity = math.abs(one - 1) < 1e-9

    override def changeToInternal(ds: Array[Double]) { 
      if (math.abs(one-1) > 1e-9) { var i = 0; while (i < ds.length) { ds(i) = ds(i) * one; i +=1 } }
    }
    override def changeToInternal(fs: Array[Float]) { 
      if (math.abs(one-1) > 1e-9) { var i = 0; while (i < fs.length) { fs(i) = (fs(i) * one).toFloat; i += 1 } }
    }

    override def changeToExternal(ds: Array[Double]) { 
      if (math.abs(one-1) > 1e-9) { var i = 0; while (i < ds.length) { ds(i) = ds(i) / one; i +=1 } }
    }
    override def changeToExternal(fs: Array[Float]) { 
      if (math.abs(one-1) > 1e-9) { var i = 0; while (i < fs.length) { fs(i) = (fs(i) / one).toFloat; i += 1 } }
    }

    def *(lc: LinearConvert) =
      new LinearConvert(s"($name) * (${lc.name})", one * lc.one, dDim + lc.dDim, tDim + lc.tDim, aDim + lc.aDim)
    def /(lc: LinearConvert) = 
      new LinearConvert(s"($name) / (${lc.name})", one / lc.one, dDim - lc.dDim, tDim - lc.tDim, aDim - lc.aDim)
    def ^(n: Int) =
      new LinearConvert(s"($name)^$n", math.pow(one, n), dDim*n, tDim*n, aDim*n)

    def e(n: Int) = copy(one = if (n >= 0) one * ("1e"+n.toString).toDouble else one / ("1e"+(-n).toString).toDouble)
  }

  object Standard {
    val millimeter = LinearConvert("mm", 1, 1, 0, 0)
    val meter = LinearConvert("m", 1e3, 1, 0, 0)
    val micron = LinearConvert("um", 1e-3, 1, 0, 0)
    val inch = LinearConvert("in", 25.4, 1, 0, 0)
    val foot = LinearConvert("ft", 304.8, 1, 0, 0)
    val second = LinearConvert("s", 1, 0, 1, 0)
    val minute = LinearConvert("min", 60, 0, 1, 0)
    val hour = LinearConvert("hr", 3600, 0, 1, 0)
    val day = LinearConvert("day", 86400, 0, 1, 0)
    val radian = LinearConvert("radian", 1, 0, 0, 1)
    val degree = LinearConvert("degree", 180/math.Pi, 0, 0, 1)
    val one = LinearConvert("", 1, 0, 0, 0)
    def scalar(x: Double) = LinearConvert(x.toString, x, 0, 0, 0)
    val percent = LinearConvert("%", 0.01, 0, 0, 0)
    val celsius = IdentityConvert("C")
    val kelvin = new Convert { def name = "K"; def toInternal(x: Double) = x - 273.15; def toExternal(x: Double) = x + 273.15 }
    val fahrenheit = new Convert { def name = "F"; def toInternal(x: Double) = ((x-32)*5)/9; def toExternal(x: Double) = (x*9)/5 + 32 }    
  }

  object ConvertParser {
    import Standard._
    import fastparse.all._
    val W = P(CharsWhile(_.isWhitespace).?)
    val D = CharIn("0123456789")
    val Dp = CharIn("123456789")
    val PositiveDouble = P(
      ( P("0." ~ "0".rep ~ Dp ~ D.rep) | 
        P(Dp ~ D.rep ~ ("." ~ D.rep(1)).?)
      ) ~ 
      (CharIn("eE") ~ CharIn("+-").? ~ D.rep(1)).?
    ).!.map(_.toDouble)
    // Be SURE to put the longer version first or it will not be matched!
    val LongLinBase = 
      P("meters" | "meter" | "metres" | "metre").map(_ => meter) |
      P("inches" | "inch").map(_ => inch) |
      P("foot" | "feet").map(_ => foot) |
      P("seconds" | "second").map(_ => second) |
      P("minutes" | "minute").map(_ => minute) |
      P("hours" | "hour").map(_ => hour) |
      P("days" | "day").map(_ => day) |
      P("radians" | "radian").map(_ => radian) |
      P("degrees" | "degree").map(_ => degree)
    val LongPrefix =
      P("centi").map(_ => -2) |
      P("milli").map(_ => -3) |
      P("micro").map(_ => -6) |
      P("nano").map(_ => -9) |
      P("kilo").map(_ => 3) |
      P("mega").map(_ => 6) |
      P("giga").map(_ => 9)
    val LongLin =
      P("micron" | "microns").map(_ => micron) |
      (LongPrefix ~ LongLinBase).map{ case (a,b) => b e a } |
      LongLinBase
    val ShortLinBase =
      P("rad").map(_ => radian) |
      P("deg").map(_ => degree) |  // Must go above "d"!
      P("s").map(_ => second) |
      P("min").map(_ => minute) |  // Must go above "m"!
      P("hr" | "h").map(_ => hour) |
      P("d").map(_ => day) |
      P("m").map(_ => meter) |
      P("in").map(_ => inch) |
      P("ft").map(_ => foot)
    val ShortPrefix =
      P("c").map(_ => -2) |
      P("m").map(_ => -3) |
      P("u" | "\u00B5" | "\u03BC").map(_ => -6) |
      P("n").map(_ => -9) |
      P("k").map(_ => 3) |
      P("M").map(_ => 6) |
      P("G").map(_ => 9)
    val ShortLin =
      P("mm").map(_ => millimeter) |
      P("%").map(_ => percent) |
      (ShortPrefix ~ ShortLinBase).map{ case (a,b) => b e a } |
      ShortLinBase
    type PLC = Parser[LinearConvert]
    val OneLin: PLC = P(PositiveDouble.map(x => if (x==1) one else scalar(x)) | LongLin | ShortLin)
    val ExpLin: PLC = P((ParLin ~ W ~ "^" ~ W ~ P("-".? ~ CharsWhile(_.isDigit)).!.map(_.toDouble.toInt)).map{ case (a,b) => a ^ b })
    val PrecA: PLC  = P(ExpLin | ParLin)
    val MulLin: PLC = P(((PrecA ~ W ~ PrecB) | (PrecA ~ W ~ "*" ~ W ~ PrecB)).map{ case (a,b) => a * b })
    val PrecB: PLC  = P(MulLin | PrecA)
    val DivLin: PLC = P((PrecB ~ W ~ "/" ~ W ~ PrecB).map{ case (a,b) => a / b })
    val PrecC: PLC  = P(DivLin | PrecB)
    val ParLin: PLC = P("(" ~ W ~ PrecC ~ W ~ ")") | OneLin
    val NonLin =
      P("celsius" | "Celsius" | "centigrade" | "Centigrade" | "C").map(_ => celsius) |
      P("fahrenheit" | "Fahrenheit" | "F").map(_ => fahrenheit) |
      P("kelvin" | "Kelvin" | "K").map(_ => kelvin)
    val Cvt = (PrecC ~ End) | (NonLin ~ End)

    def apply(s: String): Option[Convert] = Cvt.parse(s) match {
      case Parsed.Success(x, _) => Some(x)
      case _ => None
    }
  }
}

package object units {
  def parseUnit(s: String): Option[Convert] = ConvertParser(s) match {
    case Some(lc: LinearConvert) => Some(lc.copy(name = s))
    case x                       => x
  }

  implicit val parseConvertFromJson = new FromJson[Convert] {
    def parse(j: Json): Either[JastError, Convert] = j match {
      case Json.Str(text) => parseUnit(text) match {
        case Some(c) => Right(c)
        case None => Left(JastError("Could not parse unit "+text))
      }
      case _ => Left(JastError("No unit conversion: value is not a string"))
    }
  }

  implicit class FloatIntoUnits(private val underlying: Float) extends AnyVal {
    def into(u: Convert): Float = (u toExternal underlying).toFloat
    def from(u: Convert): Float = (u toInternal underlying).toFloat
  }
  implicit class ArrayFloatIntoConvert(val underlying: Array[Float]) extends AnyVal {
    def changeInto(u: Convert): underlying.type = { u changeToExternal underlying; underlying }
    def changeFrom(u: Convert): underlying.type = { u changeToInternal underlying; underlying }
  }
  implicit class ArrayArrayFloatIntoConvert(val underlying: Array[Array[Float]]) extends AnyVal {
    def changeInto(u: Convert): underlying.type = { u changeToExternal underlying; underlying }
    def changeFrom(u: Convert): underlying.type = { u changeToInternal underlying; underlying }
  }
  implicit class DoubleIntoConvert(private val underlying: Double) extends AnyVal {
    def into(u: Convert): Double = u toExternal underlying
    def from(u: Convert): Double = u toInternal underlying
  }
  implicit class ArrayDoubleIntoConvert(val underlying: Array[Double]) extends AnyVal {
    def changeInto(u: Convert): underlying.type = { u changeToExternal underlying; underlying }
    def changeFrom(u: Convert): underlying.type = { u changeToInternal underlying; underlying }
  }
  implicit class ArrayArrayDoubleIntoConvert(val underlying: Array[Array[Double]]) extends AnyVal {
    def changeInto(u: Convert): underlying.type = { u changeToExternal underlying; underlying }
    def changeFrom(u: Convert): underlying.type = { u changeToInternal underlying; underlying }
  }
}

File: src/julia/src/Units.jl
## This file is part of Open Worm's Tracker Commons project and is distributed under the MIT license.
## Contents copyright (c) 2016 by Rex Kerr, Calico Life Sciences, and Open Worm.

#################################################
# Implementation of Tracker Commons recommended #
# unit conversions for WCON format              #
#################################################

si_prefix_map = Dict(
    "c" => 1e-2,
    "centi" => 1e-2,
    "m" => 1e-3,
    "milli" => 1e-3,
    "u" => 1e-6,
    "\u00B5" => 1e-6,
    "\u03BC" => 1e-6,
    "micro" => 1e-6,
    "n" => 1e-9,
    "nano" => 1e-9,
    "k" => 1e3,
    "kilo" => 1e3,
    "M" => 1e6,
    "mega" => 1e6,
    "G" => 1e9,
    "giga" => 1e9
)

nonscaled_units_to_internal = Dict(
    "C" => x :: Float64 -> x,
    "K" => x :: Float64 -> x - 273.15,
    "F" => x :: Float64 -> ((x - 32)*5.0)/9.0
)

nonscaled_units_to_external = Dict(
    "C" => x :: Float64 -> x,
    "K" => x :: Float64 -> x + 273.15,
    "F" => x :: Float64 -> (x*9.0)/5.0 + 32
)

long_units_list = collect(Dict(
    "meter" => 1e3,
    "metre" => 1e3,
    "meters" => 1e3,
    "metres" => 1e3,
    "micron" => 1e-3,
    "microns" => 1e-3,
    "inch" => 25.4,
    "inches"=> 25.4,
    "second" => 1.0,
    "seconds" => 1.0,
    "minute" => 60.0,
    "minutes" => 60.0,
    "hour" => 3600.0,
    "hours" => 3600.0,
    "day" => 86400.0,
    "days" => 86400.0
))

abbrev_units_list = collect(Dict(
    "1" => 1.0,
    "%" => 0.01,
    "m" => 1e3,
    "in" => 25.4,
    "s" => 1.0,
    "min" => 60.0,
    "h" => 3600.0,
    "hr" => 3600.0,
    "d" => 86400.0
))

function make_nonscaled_converters(input :: AbstractString)
    result :: Nullable{Tuple{Function, Function}} = Nullable{Tuple{Function, Function}}()
    if haskey(nonscaled_units_to_internal, input) && haskey(nonscaled_units_to_external)
        result = Nullable(nonscaled_units_to_internal(input), nonscaled_units_to_external(input))
    end
    return result
end

function make_simple_scale(input :: AbstractString)
    result :: Float64 = NaN
    tidy = strip(input)
    longi = findfirst(x -> endswith(tidy, first(x)), long_units_list)
    if longi > 0
        k, v = long_units_list[longi]
        if tidy == k result = v
        else
            pre = tidy[1:(length(tidy) - length(k))]
            if haskey(si_prefix_map, pre) && length(pre) > 1
                result = v * si_prefix_map[pre]
            end
        end
    else
        shorti = findfirst(x -> endswith(input,first(x)), abbrev_units_list)
        if shorti > 0
            k, v = abbrev_units_list[shorti]
            if tidy == k result = v
            elseif k != "1" && k != "%"
                pre = tidy[1:(length(tidy)-length(k))]
                if haskey(si_prefix_map, pre) && length(pre) == 1
                    result = v * si_prefix_map[pre]
                end
            end
        end
    end
end

function make_complex_scale(input :: AbstractString)
    result :: Float64 = NaN
    div = findfirst(input, '/')
    if (div > 0)
        result = make_complex_scale(input[1:(div-1)]) / make_complex_scale(input[(div+1):end])
    else
        mul = findfirst(input, '*')
        if (mul > 0)
            result = make_complex_scale(input[1:(mul-1)]) * make_complex_scale(input[(div+1):end])
        else
            pwr = findfirst(input, '^')
            if (pwr > 0)
                result = make_simple_scale(input[1:(pwr-1)]) ^ (try parse(Int8, strip(input[(pwr+1):end])) catch; NaN end)
            else
                result = make_simple_scale(input)
            end
        end
    end
    return result
end

function make_complex_converters(input :: AbstractString)
    result :: Nullable{Tuple{Function, Function}} = Nullable{Tuple{Function, Function}}()
    sks = make_complex_scale(input)
    if !isnan(sks)
        result = Nullable((x :: Float64 -> x * sks, x :: Float64 -> x / sks))
    end
    return result
end

function make_any_converters(input :: AbstractString)
    result :: Nullable{Tuple{Function, Function}} = Nullable{Tuple{Function, Function}}()
    nsc = make_nonscaled_converters(strip(input))
    result = nsc.isnull ? make_complex_converters(input) : Nullable(nsc.value)
    return result
end

type UnitConverter
    description :: AbstractString
    to_internal :: Function
    to_external :: Function
end

type KnownUnits
    mapper :: Dict{AbstractString, UnitConverter}
    custom :: Dict{AbstractString, Any}
end

function minimal_known_units()
    convert(
      KnownUnits,
      parsed_json_to_units(Dict{AbstractString, Any}("t" => "s", "x" => "mm", "y" => "mm"))
    )
end

function convert_for_json(ku :: KnownUnits)
    result = Dict{AbstractString, Any}()
    for (k,v) in ku.mapper
        if k == "cx"
            if ku.mapper["x"].description != v.description result[k] = v.description end
        elseif k == "cy"
            if ku.mapper["y"].description != v.description result[k] = v.description end
        elseif k == "ox"
            if ku.mapper["x"].description != v.description result[k] = v.description end
        elseif k == "oy"
            if ku.mapper["y"].description != v.description result[k] = v.description end
        else
            result[k] = v.description
        end
    end
    if length(ku.custom) > 0
        for (k,v) in ku.custom 
            result[k] = v
        end
    end
    return result
end

function parsed_json_to_units(x :: Dict{AbstractString, Any})
    result :: Union{AbstractString, KnownUnits} = "Units block parsing failed"
    m = Dict{AbstractString, UnitConverter}()
    c = Dict{AbstractString, Any}()
    for (k,v) in x
        if startswith(k, "@")
            c[k] = v
        else
            if isa(v, AbstractString)
                s = convert(AbstractString, v)
                u = make_any_converters(s)
                if u.isnull
                    result = string("Units of ",k," cannot be parsed: ", s)
                    return result
                elseif haskey(m, k)
                    result = string("Units of ",k," should not be given more than once")
                    return result
                else
                    (fwd,bkw) = u.value
                    m[k] = UnitConverter(s, fwd, bkw)
                end
            else
                result = string("Units of ",k," should be given as a string")
                return result
            end
        end
    end
    if !haskey(m,"t") || !haskey(m,"x") || !haskey(m, "y")
        result = string("Units must be specified for all of t, x, and y")
    else
        result = KnownUnits(m,c)
    end
    return result
end

function floatize_any_array(a :: Array{})
    if isa(a, Array{Float64, 1})
        return a
    end
    result = fill(NaN, length(a))
    for i in 1:length(a)
        x = a[i]
        if isa(x, Number)
            result[i] = convert(Float64, x)
        elseif isa(x, Void)
            result[i] = NaN
        else
            return a
        end
    end
    return result
end

function internalize_parsed_json!(j :: Any, ku :: KnownUnits, uc :: Nullable{UnitConverter})
    if isa(j, Dict)
        d = convert(Dict,j)
        for (k,v) in d
            if k != "settings"
                uk = haskey(ku.mapper, k) ? Nullable(ku.mapper[k]) : uc
                if isa(v, Number) && !uk.isnull
                    d[k] = uk.value.to_internal(convert(Float64, v))
                else
                    x =
                        if isa(v, Array{})
                            if isa(v, Array{Float64, 1}) convert(Array{Float64, 1}, j)
                            else
                                y = floatize_any_array(convert(Array{}, v))
                                d[k] = y
                                y
                            end
                        else
                            v
                        end
                    internalize_parsed_json!(x, ku, uk)
                end
            end
        end
    elseif isa(j, Array{})
        ja = convert(Array{}, j)
        for i in 1:length(ja)
            ji = ja[i]
            if isa(ji, Number) && !uc.isnull
                ja[i] = uc.value.to_internal(convert(Float64, ji))
            else
                if isa(ji, Array{})
                    x = floatize_any_array(ji)
                    if isa(x, Array{Float64, 1})
                        ja[i] = x
                        if !uc.isnull
                            ys = convert(Array{Float64, 1}, x)
                            fn = uc.value.to_internal
                            for i in 1:length(ys)
                                ys[i] = fn(ys[i])
                            end
                        end
                    else
                        internalize_parsed_json!(ji, ku, uc)
                    end
                else
                    internalize_parsed_json!(ji, ku, uc)
                end
            end
        end
    end
end

function externalize_jsonable!(j :: Any, ku :: KnownUnits, uc :: Nullable{UnitConverter})
    if isa(j, Dict)
        d = convert(Dict, j)
        for (k,v) in d
            if k != "settings"
                uk = haskey(ku.mapper, k) ? Nullable(ku.mapper[k]) : uc
                if isa(v, Float64) && ! uk.isnull
                    d[k] = uk.value.to_external(convert(Float64, v))
                else
                    externalize_jsonable!(v, ku, uk)
                end
            end
        end
    elseif isa(j, Array{})
        if isa(j, Array{Float64, 1}) && !uc.isnull
            fn = uc.value.to_external
            jf = convert(Array{Float64, 1}, j)
            for i in 1:length(jf)
                jf[i] = fn(jf[i])
            end
        else
            ja = convert(Array{}, j)
            if uc.isnull
                for v in ja externalize_jsonable!(v, ku, uc) end
            else
                for i in 1:length(ja)
                    if isa(ja[i], Float64) ja[i] = fn(convert(Float64,ja[i]))
                    else
                        externalize_jsonable!(ja[i], ku, uc)
                    end
                end
            end
        end
    end
end
